src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,368 @@
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mHomeLocation.mFloorNumber = aFloorNumber;
+    return *this;
+}
+
+EcosystemLocationStruct::Builder & EcosystemLocationStruct::Builder::SetAreaTypeTag(std::optional<AreaTypeTag> aAreaTypeTag)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mHomeLocation.mAreaType = aAreaTypeTag;
+    return *this;
+}
+
+EcosystemLocationStruct::Builder & EcosystemLocationStruct::Builder::SetHomeLocationLastEdit(uint64_t aHomeLocationLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mHomeLocationLastEditEpochUs = aHomeLocationLastEditEpochUs;
+    return *this;
+}
+
+std::unique_ptr<EcosystemLocationStruct> EcosystemLocationStruct::Builder::Build()
+{
+    bool HomeLocationNameIsInvalid =
+        mHomeLocation.mLocationName.empty() || mHomeLocation.mLocationName.size() > kHomeLocationNameMaxSize;
+    if (mIsAlreadyBuilt || HomeLocationNameIsInvalid)
+    {
+        return nullptr;
+    }
+
+    // std::make_unique does not have access to private constructor we workaround with using new
+    std::unique_ptr<EcosystemLocationStruct> ret{ new EcosystemLocationStruct(std::move(mHomeLocation),
+                                                                              mHomeLocationLastEditEpochUs) };
+    mIsAlreadyBuilt = true;
+    return ret;
+}
+
+CHIP_ERROR EcosystemLocationStruct::Encode(const AttributeValueEncoder::ListEncodeHelper & aEncoder,
+                                           const std::string & aUniqueLocationId, const FabricIndex & aFabricIndex)
+{
+    Structs::EcosystemLocationStruct::Type locationStruct;
+    VerifyOrDie(!aUniqueLocationId.empty());
+    locationStruct.uniqueLocationID     = CharSpan(aUniqueLocationId.c_str(), aUniqueLocationId.size());
+    locationStruct.homeLocation         = GetEncodableHomeLocationStruct(mHomeLocation);
+    locationStruct.homeLocationLastEdit = mHomeLocationLastEditEpochUs;
+
+    // TODO(#33223) this is a hack, use mFabricIndex when it exists. Additionally checking fabric
+    // index matches should happen at the top of this method to prevent building the encodable
+    // device struct when we know it is not intended for the accessing fabric.
+    locationStruct.SetFabricIndex(aFabricIndex);
+    return aEncoder.Encode(locationStruct);
+}
+
+EcosystemInformationServer EcosystemInformationServer::mInstance;
+
+EcosystemInformationServer & EcosystemInformationServer::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR EcosystemInformationServer::AddDeviceInfo(EndpointId aEndpoint, std::unique_ptr<EcosystemDeviceStruct> aDevice)
+{
+    VerifyOrReturnError(aDevice, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError((aEndpoint != kRootEndpointId && aEndpoint != kInvalidEndpointId), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto & deviceInfo = mDevicesMap[aEndpoint];
+    VerifyOrReturnError((deviceInfo.mDeviceDirectory.size() >= kDeviceDirectoryMaxSize), CHIP_ERROR_NO_MEMORY);
```

Comment:
this verify seems reversed ... should we verify that the size is smaller than the maximum size and otherwise return an error?

Same below.

Reply:
Good catch.

This comment is also now left open until I confirm with my local setup that my latest updates and check work as expected

Reply:
Confirmed I was able to add and read values from Device and Location directory